modules/memberships/class-jetpack-memberships.php

```diff
@@ -0,0 +1,283 @@
+	 *
+	 * @var string
+	 */
+	static private $version = JETPACK__VERSION;
+	/**
+	 * Our CPT type for the product (plan).
+	 *
+	 * @var string
+	 */
+	static public $post_type_plan = 'jp_mem_plan';
+	/**
+	 * @var string Option that will store currently set up account (Stripe etc) id for memberships
+	 */
+	static public $connected_account_id_option_name = 'jetpack-memberships-connected-account-id';
+	/**
+	 * Button block type to use.
+	 *
+	 * @var string
+	 */
+	static private $button_block_name = 'membership-button';
+	/**
+	 * Classic singleton pattern
+	 *
+	 * @var Jetpack_Memberships
+	 */
+	private static $instance;
+
+	/**
+	 * Jetpack_Memberships constructor.
+	 */
+	private function __construct() {}
+
+	/**
+	 * The actual constructor initializing the object.
+	 *
+	 * @return Jetpack_Memberships
+	 */
+	public static function get_instance() {
+		if ( ! self::$instance ) {
+			self::$instance = new self();
+			self::$instance->register_init_hook();
+		}
+
+		return self::$instance;
+	}
+	/**
+	 * Get the map that defines the shape of CPT post. keys are names of fields and
+	 * 'meta' is the name of actual WP post meta field that corresponds.
+	 *
+	 * @return array
+	 */
+	private static function get_plan_property_mapping() {
+		$meta_prefix = 'mem_';
+		$properties  = array(
+			'price'    => array(
+				'meta' => $meta_prefix . 'price',
+			),
+			'currency' => array(
+				'meta' => $meta_prefix . 'currency',
+			),
+		);
+		return $properties;
+	}
+
+	/**
+	 * Transform WP CPT post into array representing a memberships product.
+	 *
+	 * @param WP_Post $product_post - CPT representing the product.
+	 * @return array
+	 */
+	public static function product_post_to_array( $product_post ) {
+		$data    = array();
+		$mapping = self::get_plan_property_mapping();
+		foreach ( $mapping as $key => $map ) {
+			$data[ $key ] = get_post_meta( $product_post->ID, $map['meta'], true );
+		}
+		$data['title']       = $product_post->post_title;
+		$data['description'] = $product_post->post_content;
+		$data['id']          = $product_post->ID;
+		return $data;
+	}
+
+	/**
+	 * Inits further hooks on init hook.
+	 */
+	private function register_init_hook() {
+		add_action( 'init', array( $this, 'init_hook_action' ) );
+	}
+
+	/**
+	 * Actual hooks initializing on init.
+	 */
+	public function init_hook_action() {
+		add_filter( 'rest_api_allowed_post_types', array( $this, 'allow_rest_api_types' ) );
+		add_filter( 'jetpack_sync_post_meta_whitelist', array( $this, 'allow_sync_post_meta' ) );
+		$this->setup_cpts();
+	}
+
+	/**
+	 * Sets up the custom post types for the module.
+	 */
+	private function setup_cpts() {
+		/*
+		 * PLAN data structure.
+		 */
+		$capabilities = array(
+			'edit_post'          => 'edit_posts',
+			'read_post'          => 'read_private_posts',
+			'delete_post'        => 'delete_posts',
+			'edit_posts'         => 'edit_posts',
+			'edit_others_posts'  => 'edit_others_posts',
+			'publish_posts'      => 'publish_posts',
```

Comment:
Does this capability make it so that the block doesn't work for contributor users? If so, does the editor reflect that graciously without failure?

Reply:
Since these are changed by custom API and access methods, I opted out for restricting user capabilities to change them.
It all graciously handles these cases.